craft_application/launchpad/models/recipe.py

```diff
@@ -0,0 +1,351 @@
+# pyright: reportOptionalCall=false
+# pyright: reportOptionalIterable=false
+
+from __future__ import annotations
+
+import enum
+import time
+import typing
+from collections.abc import Collection, Iterable
+from typing import TYPE_CHECKING
+
+import lazr.restfulclient.errors  # type: ignore[import-untyped]
+from typing_extensions import Any, Self, TypedDict, override
+
+from .. import errors, util
+from . import build
+from .base import LaunchpadObject, Pocket
+
+if TYPE_CHECKING:
+    from .. import Launchpad
+
+
+class RecipeType(enum.Enum):
+    """The type of recipe."""
+
+    SNAP = "snap"
+    CHARM = "charm_recipe"
+
+
+class BuildChannels(TypedDict, total=False):
+    """Typed dictionary for the snap channels for recipe builds."""
+
+    # Snaps and charms
+    core: str
+    core18: str
+    core20: str
+    core22: str
+    core24: str
+
+    charmcraft: str  # Charms only
+
+    # Snaps only
+    snapcraft: str
+    snapd: str
+
+
+class _BaseRecipe(LaunchpadObject):
+    name: str
+    """The name of the recipe."""
+    owner_name: str
+    """The name of the owner."""
+
+    _resource_types = RecipeType
+    _attr_map = {"owner_name": "owner.name"}
+
+    def __repr__(self) -> str:
+        return f"{self.__class__.__name__}.get(name={self.name!r}, owner={self.owner_name!r})"
+
+    @staticmethod
+    def _fill_repo_info(
+        kwargs: dict[str, Any],
+        *,
+        git_ref: str | None = None,
+        bzr_branch: str | None = None,
+    ) -> None:
+        """Conditionally fill source repository info into keyword arguments."""
+        if (git_ref and bzr_branch) or (not git_ref and not bzr_branch):
+            raise ValueError(
+                "A snap recipe must refer to a git repository or a bazaar branch, "
+                "but not both.",
+            )
+        if git_ref:
+            kwargs["git_ref"] = git_ref
+        if bzr_branch:
+            kwargs["branch"] = bzr_branch
+
+    def get_builds(self) -> Collection[build.Build]:
+        """Get the existing builds for a Recipe."""
+        return [
+            build.Build(self._lp, b)
+            for b in self._obj.builds  # pyright: ignore[reportGeneralTypeIssues]
+        ]
+
+    def build(
+        self,
+        archive: str = "/ubuntu/+archive/primary",
+        pocket: Pocket = Pocket.UPDATES,
+        channels: BuildChannels | None = None,
+        deadline: int | None = None,
+    ) -> Collection[build.Build]:
+        """Create a new set of builds for this recipe."""
+        request_build_kwargs: dict[str, Any] = {
+            "archive": archive,
+            "pocket": pocket.value,
+        }
+        if channels:
+            request_build_kwargs["channels"] = channels
+        build_request = self._obj.requestBuilds(**request_build_kwargs)
+        sleep_time = 0.5
+        while build_request.status == "Pending":
+            # Check to see if we've run out of time.
+            if deadline is not None and time.monotonic_ns() >= deadline:
```

Comment:
This needs to use nanoseconds?

Reply:
No, but `time.monotonic()` is basically just `time.monotonic_ns() / 10**9` and I prefer working with integers.